Targets: Support OpenNaze32 (compatible with Naze32)

mlyle opened the pull request:
This provides support for the Naze family of boards. The -rev4 schematic was released by the
original author Timecop under a CC-BY-SA-NC 4.0 license. This was subsequently recreated and
a new layout (OpenNaze) was developed by @readerror67. The design files for this (both
schematic and gerbers are included in the targets/naze/hw directory under CC-BY-SA-NC 4.0) as well
 as the original schematic released by TimeCop.

It is worth noting there is a subsequent -rev5 design that has been released which has two
relevant changes: a 12MHz crystal and moving the MPU-6050 IRQ from PB13 to PC13.
This code _does_ support this hardware as well, although there is no open hardware 
implementation. This seems like a minor enough (and well documented) change that it is
worth including.

BrainFPV commented:
I was under the impression that Naze32 is closed HW, so can't be included in TL.

peabody124 commented:
@mlyle you should change this to support OpenNaze32 and make sure to include the hardware files (provided ReadError is ok and licenses those files appropriately)

peabody124 commented:
Another thing on top of the hardware files (appropriately licensed with clean IP) that should be done. There should be a new board ID assigned to this and the GCS informed about it. My code just made it fake out that it was a CC3D which probably isn't the best way to go.

FTW I also have some old serial STM32 boot loader code ported to Qt I used for the ESCs we could use if designed. That is definitely way beyond the scope of initial support.

tracernz commented:
I have added a board ID and GCS plugin in my branch already. https://github.com/tracernz/TauLabs/tree/naze32-next
I'll just need to do something about the bootloader..

BrainFPV commented on flight/Modules/Attitude/naze32/attitude.c:
```diff
@@ -0,0 +1,730 @@
+/**
```
I assume this is in a separate file to have a lightweight attitude estimation for the Naze32. I think it would make sense to have this in a more general file, so it could be used by other targets (e.g. CC3D).

mlyle replied:
Yes.  This is very similar to the CC3D attitude code.  It's unfortunate the amount of duplication between the CC3D code, this code, and the main attitude code.  They're different because CC3D and this code talk directly to the sensors and there's some differences.  As part of target cleanup I'm going to work in this area (hopefully getting some additional functionality for CC3D in the process).

BrainFPV commented on ground/gcs/src/plugins/boards_naze/Naze.pluginspec:
```diff
@@ -0,0 +1,12 @@
+<plugin name="Naze" version="1.0.0" compatVersion="1.0.0">
+    <vendor>Tau Labs</vendor>
```
Should probably be "tracernz" (to be consistent with description below)

BrainFPV commented on flight/targets/naze32/fw/Makefile:
```diff
@@ -0,0 +1,734 @@
+ifeq ($(NAVIGATION), YES)
+MODULES += State/naze32
+MODULES += PathFollower
+CFLAGS += -DNAVIGATION
```
`Attitude` seems to be missing

BrainFPV commented on flight/targets/naze32/fw/Makefile:
```diff
@@ -0,0 +1,734 @@
+endif
+# List of mandatory modules to include
+ifeq ($(NAVIGATION), YES)
+MODULES += State/naze32
```
This module doesn't seem to be included in PR

solidgoldbomb commented:
Here is a release by the original designer @trollcop (dongs) of the Naze32 rev4 schematic that is marked (on the drawing) with a CC BY-NC-SA 3.0 license:

* https://code.google.com/p/afrodevices/downloads/detail?name=naze32_rev4.pdf

Here are the OpenNaze_rev4 docs:

* https://github.com/Brotronics/OpenNaze_rev4

The readme.md file says that the files in that directory are licensed as CC BY-NC-SA 4.0 but I can't read any of the files to see what license is written on the drawings there.  @readerror67, are any of these files a schematic for OpenNaze?  Are there any differences from Naze32 rev4?  If so, could you add a schematic PDF to your git tree so we can use it as reference?

trollcop commented:
you guys basically eat dicks :(
On Jul 10, 2015 11:09 PM, "Stacey Sheldon" <notifications@github.com> wrote:

> Here is a release by the original designer @trollcop
> <https://github.com/trollcop> (dongs) of the Naze32 rev4 schematic that
> is marked (on the drawing) with a CC BY-NC-SA 3.0 license:
>
>    -
>    https://code.google.com/p/afrodevices/downloads/detail?name=naze32_rev4.pdf
>
> Here are the OpenNaze_rev4 docs:
>
>    - https://github.com/Brotronics/OpenNaze_rev4
>
> The readme.md file says that the files in that directory are licensed as
> CC BY-NC-SA 4.0 but I can't read any of the files to see what license is
> written on the drawings there. @readerror67
> <https://github.com/readerror67>, are any of these files a schematic for
> OpenNaze? Are there any differences from Naze32 rev4? If so, could you add
> a schematic PDF to your git tree so we can use it as reference?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/TauLabs/TauLabs/pull/1711#issuecomment-120418144>.
>

readerror67 commented:
There are a few differences, one being support for MPU9150 (rather than the 6050), I also removed the HMC5883 mag since many GPS include them these days and it avoids the need to solder jumpers to enable/disable it if you are using an external one.

There is also a multiplexer on the USART1_RX that changes the input based on VBUS (USB connected -> OSD TX is disconnected, prevents OSD from talking when trying to configure the board).

The files are DipTrace which is available on Windows/Linux/OSX free.

I added a PDF:
==
https://github.com/Brotronics/OpenNaze_rev4/blob/master/OpenNaze_rev4.pdf
==

Does the license need to be embedded in to that document ?

solidgoldbomb commented:
Thanks for posting the PDF @readerror67.  Embedding the license in the doc makes the license clear even after it has been downloaded.  We typically include the schematics in our repo so it's more self-contained if the license is specifically stated within the drawing.

solidgoldbomb commented on flight/targets/naze32/target-defs.mk:
```diff
@@ -0,0 +1,10 @@
+
+# Set the cpu architecture here that matches your STM32
+# Should be one of: f1,f3,f4
+coptercontrol_cpuarch := f1
```
I think this should be ```naze32_cpuarch```.

solidgoldbomb commented on flight/targets/naze32/target-defs.mk:
```diff
@@ -0,0 +1,10 @@
+
+# Short name of this board (used to display board name in parallel builds)
+# Should be exactly 4 characters long.
+coptercontrol_short := 'naze'
```
This should be ```naze32_short```.

mlyle commented:
This is in a place where it's close to being able to be merged.  I need to carefully read the code still.  Also, all of these hardware target changes need to be committed in sequence and then fixed up (touch common lines in packaging).  #1719 should go next.

readerror67 commented:
I have added the name/license to the schematic file and PDF.
https://github.com/Brotronics/OpenNaze_rev4/blob/master/OpenNaze_rev4.pdf

Let me know if I am missing anything :)

mlyle commented:
Picked up new schematic PDF and rebased into the PR.

I'm now beginning my skim of the diff in earnest.

mlyle commented:
Fixed up the attitude stuff to be factored common-- you're right @BrainFPV that was too disgusting for initial commit even.  Also removed vestiges of NAVIGATION that can't build from CC3D so that these targets are closer together.

I ran a big sed, which had a side effect of making sure files ended with newlines.  Hope that's OK.

mlyle commented:
Fixed up because of dependency on #1719 merge.

mlyle commented:
I am pleased with this diff; needs test on CC3D & Naze32.

tracernz commented on flight/targets/naze32/board-info/board_hw_defs.c:
```diff
@@ -0,0 +1,904 @@
+void PIOS_SPI_generic_irq_handler(void);
+void DMA1_Channel4_IRQHandler() __attribute__ ((alias ("PIOS_SPI_generic_irq_handler")));
+void DMA1_Channel5_IRQHandler() __attribute__ ((alias ("PIOS_SPI_generic_irq_handler")));
+static const struct pios_spi_cfg pios_spi_generic_cfg = {
```
This SPI isn't routed on either the Naze32 rev. 4 or OpenNaze rev. 4. I believe it is used on Naze32 rev. 5 for the SPI flash chip (M25P16) but we don't support that. We also don't have PIOS_INCLUDE_SPI in pios_config.h. Maybe we should just remove SPI code?

mlyle replied:
If it's useful work towards supporting the flash chip, which it looks like it is, and doesn't break anything... I'd say we keep it.  Having a low-end platform that can log uavos would be awesome.

tracernz commented on flight/targets/naze32/link_STM32103CB_Naze32_sections.ld:
```diff
@@ -0,0 +1,173 @@
+     */
+    .fw_version_blob :
+    {
+        KEEP(*(.fw_version_blob))
```
I can't see how this would be linked? The firmware info template isn't parsed and compiled until after the elf has been linked.

mlyle replied:
It's incomplete-- we don't use the firmware info template on Naze and can't read firmware versions, so we'll have to do it differently here, I think.

tracernz replied:
Yep, something for the todo list when we figure out how to load the flash from GCS. :+1:

peabody124 replied:
i have code in my ESC branch that is a Qt plugin that implements the stm32flash stuff. With the ESC i had code in the boot loader that would allow it to reset itself into DFU mode.

mlyle commented:
Verified functional on original CC.

mlyle commented:
Verified functional (but not flown) on emax skyline32.

tracernz commented:
I have tested basic functionality on both Naze32 (rev 5) and CC3D, but not flown either. The attitude module appears to be functioning normally on both.

peabody124 commented:
any flight tests on this yet, then?

mlyle commented:
Just on CC where there's change (to confirm no regression).  @tracernz may be able to fly it soon but it needs review in any case.

SystemsGuy commented:
Building now to flight test on a NazeV4, sparky and CC3D.

tracernz commented:
This is not ready for merge. I'm having an issue with the board (Naze32 rev. 5) locking up whenever I connect my PPM receiver. I'm currently looking into it.

DTFUHF commented:
I raised PIOS_MANUAL_STACK_SIZE in pios_config.h to 700 matching cc3d settings and the PPM problem went away.

mlyle commented:
When home I will compare stack sizes between cc3d and this and take the cc3d values.

SystemsGuy commented:
Same results as tracernz.  I'll try changing the stack size later tonight and see if I can get in the air.

mlyle commented:
I just updated the pr, so you should be able to pick up as is (merged tracernz changes)

Sent from my iPhone

> On Jul 19, 2015, at 2:51 PM, SystemsGuy <notifications@github.com> wrote:
> 
> Same results as tracernz. I'll try changing the stack size later tonight and see if I can get in the air.
> 
> —
> Reply to this email directly or view it on GitHub.
>

mlyle commented:
http://jar.lyle.org/~mlyle/TauLabs-20150719-c4c4ad58-install.exe   is a Windows installer for people with difficulty building-- it's this pull req rebased to next.

SystemsGuy commented:
Changing stack size seems to have done the trick - limited testing due to lack of sunlight, but it flew inside for a bit.

tracernz commented:
Since fixing the stack sizes (5b98d5a) I have test flown this on both Naze32 rev. 5 and CC3D. Flew with a PPM RX in acro and horizon mode with full throttle punchouts etc. on a 4S powered 250 without issue. I'm happy with this now. :+1: I have also now reviewed all of the code and am satisfied. There is plenty of work to do before the target is fully supported but this PR will get us flyable.

trollcop commented:
just for the record, I hate you all
On Jul 20, 2015 1:21 PM, "tracernz" <notifications@github.com> wrote:

> Since fixing the stack sizes (5b98d5a
> <https://github.com/TauLabs/TauLabs/commit/5b98d5ae5620186e88923661bbe1ebc869f1e17a>)
> I have test flown this on both Naze32 rev. 5 and CC3D. Flew with a PPM RX
> in acro and horizon mode with full throttle punchouts etc. on a 4S powered
> 250 without issue. I'm happy with this now. [image: :+1:] I have also now
> reviewed all of the code and am satisfied. There is plenty of work to do
> before the target is fully supported but this PR will get us flyable.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/TauLabs/TauLabs/pull/1711#issuecomment-122756837>.
>

creyc commented:
Was able to hover a small quad indoors using July 19th build on a Naze32 acro board using PPM RX.  Still needs true flight test and tuning outdoors tomorrow.

mlyle commented:
New commit ground-tested, greatly simplifies development (avoids bootjumping).

peabody124 commented on flight/Modules/Attitude/smallf1/attitude.c:
```diff
@@ -73,10 +72,17 @@ static SensorSettingsData sensorSettings;
 
 static float gyro_correct_int[3] = {0,0,0};
+
+#ifdef COPTERCONTROL
```
are we using a SMALLF1 and COPTERCONTROL define or is this a typo?

mlyle replied:
We're using both.  SMALLF1 means we're on a resource limited platform.  COPTERCONTROL is used within the smallf1 attitude module to indicate that we want the few lines particular to CC/CC3D that don't apply to Naze32.

peabody124 commented on flight/targets/coptercontrol/board-info/cmsis_system.c:
```diff
@@ -201,6 +201,10 @@ static void SetSysClock(void);
   */
 
+// this is required for the system_rcc code because
+// of the clock rate switching in naze32
```
typo. this is for coptercontrol (or in that directory).

mlyle replied:
It's actually needed so that the hse_value can be present on all smallf1 targets, since naze32 needs it.

peabody124 commented on flight/targets/coptercontrol/fw/Makefile:
```diff
@@ -53,8 +53,6 @@ ENABLE_AUX_UART ?= NO
 
 # @endgroup Compile Options
 
-NAVIGATION ?= NO
```
good drop

peabody124 commented on flight/targets/coptercontrol/fw/pios_config.h:
```diff
@@ -119,11 +119,8 @@
 #define COPTERCONTROL
-// Conditional related to making CC run navigation
-#if defined(NAVGIATION)
-#define PIOS_GPS_PROVIDES_AIRSPEED
```
this variable is still used in GPS.c -- i think we can trash those

Also seems to be used in two files in Libraries/StateEstimationFilters -- i think that directory can be trashed

There is useful stuff there but we can bring it back if needed in the future.

peabody124 commented on flight/targets/naze32/board-info/board_hw_defs.c:
```diff
@@ -0,0 +1,904 @@
+ ******************************************************************************
+ * @addtogroup TauLabsTargets Tau Labs Targets
+ * @{
+ * @addtogroup CopterControl OpenPilot naze support files
```
Probably should create a OpenNaze32 oxygen group and use that throughout these files.

peabody124 commented on flight/targets/naze32/board-info/board_hw_defs.c:
```diff
@@ -0,0 +1,904 @@
+ *
+ * @file       board_hw_defs.c 
+ * @author     The OpenPilot Team, http://www.openpilot.org Copyright (C) 2011.
+ * @author     Tau Labs, http://taulabs.org, Copyright (C) 2012-2013
```
copyright date.

peabody124 commented on flight/targets/naze32/fw/Makefile:
```diff
@@ -0,0 +1,720 @@
+###############################################################################
+# @file       Makefile
+# @author     The OpenPilot Team, http://www.openpilot.org, Copyright (C) 2009.
+# @author     Tau Labs, http://taulabs.org, Copyright (C) 2013
```
copyright. i'll stop tagging all these :)

peabody124 commented on flight/targets/naze32/fw/main.c:
```diff
@@ -0,0 +1,131 @@
+ ******************************************************************************
+ * @addtogroup TauLabsTargets Tau Labs Targets
+ * @{
+ * @addtogroup CopterControl OpenPilot naze support files
```
oxygen group. also not an open pilot board

peabody124 commented on ground/gcs/src/plugins/boards_naze/nazeplugin.cpp:
```diff
@@ -0,0 +1,65 @@
+ ******************************************************************************
+ *
+ * @file       nazeplugin.h
+ * @author     Tau Labs, http://taulabs.org, Copyright (C) 2013
```
copyright

peabody124 commented on ground/gcs/src/plugins/boards_naze/nazeplugin.h:
```diff
@@ -0,0 +1,48 @@
+ ******************************************************************************
+ *
+ * @file       nazeplugin.h
+ * @author     Tau Labs, http://taulabs.org, Copyright (C) 2013
```
copyright

peabody124 commented on package/Makefile:
```diff
@@ -45,12 +45,18 @@ CLEAN_FLIGHT := YES
 
 # Set up targets
-ALL_BOARDS        := coptercontrol pipxtreme revomini sparky sparky2 sparkybgc quanton colibri flyingf3 flyingf4
+ALL_BOARDS        := coptercontrol pipxtreme revomini sparky sparky2 sparkybgc quanton colibri flyingf3 flyingf4 naze32
```
at some point we should split all these into something that is auto-detected based on what is present in the targets directory. @solidgoldbomb if you have a brilliant makefile -foo this is one of those places that ends up creating merge conflicts when we add/remove targets

peabody124 commented:
Overall good.  few files that could be dropped (i think). a number of comments etc that need minor cleanup. code style looks good, no obvious bugs, and it has been flight tested.

peabody124 commented:
I'm happy to merge this.

The original schematic was released CC-BY-SA-NC 3.0 by the author (Timecop). 
 
A completely new layout was independently developed from this OS schematic 
by @readerror67 and is also and is also licensed CC-BY-SA 4.0 with layout files 
available for people to make their own.

The rev5 change is a single IRQ line reassignment that is well described. It would be
great to have an OSHW layout available, but I don't think this is a big enough change
to be a blocker. If we start wanting to support features on some of the boards (e.g.
Flash) that are not available in any of the open layouts, it would be prudent to discuss
this first or develop an open source schematic/layout that was compatible with those
features.

mlyle commented:
Revised per review feedback.  Thanks for the review-- I know it's tough now and I really appreciate you taking the time.

peabody124 commented on flight/targets/naze32/board-info/board-info.mk:
```diff
@@ -0,0 +1,25 @@
+BOARD_TYPE          := 0xA0
```
This needs to added to https://github.com/TauLabs/TauLabs/wiki/Development-Hardware-IDs

mlyle replied:
Done, and most of the URLs on that wiki entry corrected.

tracernz replied:
Did you mean to put 0x90 in the wiki and not 0xA0?

mlyle replied:
Nope! Fixed now, thanks.

mlyle commented:
Thanks so much for the merge!  I'm getting all the dependent PRs in the chain up now.
